api/napi/__test__/js_value_conversion.spec.ts

```diff
@@ -0,0 +1,284 @@
+  let compiler = new ComponentCompiler;
+  let definition = compiler.buildFromSource(`
+  export component App {
+    in-out property <image> image: @image-url("__test__/resources/rgb.png");
+  }
+  `, "");
+  let instance = definition?.create()!;
+
+  let slintImage = instance.getProperty("image");
+  if (t.true((slintImage instanceof ImageData))) {
+    t.deepEqual((slintImage as ImageData).width, 64);
+    t.deepEqual((slintImage as ImageData).height, 64);
+
+    Jimp.read("__test__/resources/rgb.png")
+      .then((image) => {
+        t.deepEqual(image.bitmap.data, (slintImage as ImageData).data);
+      })
+      .catch((err) => {
+        console.error(err);
+      });
+  }
+})
+
+test('get/set brush properties', (t) => {
+  let compiler = new ComponentCompiler;
+  let definition = compiler.buildFromSource(`
+  export component App {
+    in-out property <brush> black: #000000;
+    in-out property <brush> trans: transparent;
+    in-out property <brush> ref: transparent;
+  }
+  `, "");
+  let instance = definition?.create()!;
+
+  let black = instance.getProperty("black");
+
+  if (t.true((black instanceof Brush))) {
+    let blackColor = (black as Brush).color;
+    t.deepEqual(blackColor.red, 0);
+    t.deepEqual(blackColor.green, 0);
+    t.deepEqual(blackColor.blue, 0);
+  }
+
+  let transparent = instance.getProperty("trans");
+
+  if (t.true((black instanceof Brush))) {
+    t.assert((transparent as Brush).isTransparent);
+  }
+
+  let ref = Brush.fromColor(Color.fromRgb(100, 110, 120));
+  instance.setProperty("ref", ref);
+
+  let instance_ref = instance.getProperty("ref");
+
+  if (t.true((instance_ref instanceof Brush))) {
+    let ref_color = (instance_ref as Brush).color;
+    t.deepEqual(ref_color.red, 100);
+    t.deepEqual(ref_color.green, 110);
+    t.deepEqual(ref_color.blue, 120);
+  }
+})
+
+test('get/set model properties', (t) => {
+  let compiler = new ComponentCompiler;
+  let definition = compiler.buildFromSource(`
+  export component App {
+    in-out property <[string]> model: ["Florian", "Simon"];
+  }
+  `, "");
+  let instance = definition?.create()!;
+
+  let model = instance.getProperty("model");
+
+  if (t.true((model instanceof Model))) {
+    t.deepEqual((model as Model).rowCount, 2);
+  }
+})
+
+test('invoke callback', (t) => {
+  let compiler = new ComponentCompiler;
+  let definition = compiler.buildFromSource(`
+  export struct Person {
+    name: string
+  }
+  export component App {
+    callback great(string, string);
+    callback great-person(Person);
+    callback person() -> Person;
+    callback get-string() -> string;
+
+    person => {
+      {
+        name: "florian"
+      }
+    }
+
+    get-string => {
+      "string"
+    }
+  }
+  `, "");
+  let instance = definition?.create()!;
+  let speakTest;
+
+  instance.setCallback("great", (a: string, b: string) => {
+    speakTest = "hello " + a + " and " + b;
+  });
+
+  instance.invoke("great", ["simon", "florian"]);
```

Comment:
I wonder if perhaps `invoke` should not take two parameters, the name, and an array, but rather the name followed by the arguments. So `invoke(name, "simon", "florian")`. napi has the `CallContext` type for that AFAIK.

Reply:
`CallContext` is used like this:

```rust
#[js_function(1)]
pub fn call_function(ctx: CallContext) -> Result<JsNull> {
  let js_func = ctx.get::<JsFunction>(0)?;
  let js_string_hello = ctx.env.create_string("hello".as_ref())?.into_unknown();
  let js_string_world = ctx.env.create_string("world".as_ref())?.into_unknown();

  js_func.call(None, &[js_string_hello, js_string_world])?;

  ctx.env.get_null()
}
```

```ts
test('should call the function', (t) => {
  bindings.testCallFunction((arg1: string, arg2: string) => {
    t.is(`${arg1} ${arg2}`, 'hello world')
  })
})
```
It could be used by `set_callback` instead of `JsFunction` (it works both), but not for `invoke`.

Reply:
I think you're right. I just had a look and came to the same conclusion - variable arguments for a class function are just not supported out of the box right now.

I think it would be doable with a free-standing function (where we inject the typescript signature), but then we'd have to bolt that into a wrapper around the component instance, instead of using a napi class directly (for component instance).

This isn't worth the effort, so I think for now we should stick to what you implemented, invoke taking two parameters: the name, and an array of "any".